React/forms.md

```diff
@@ -2,24 +2,280 @@
+
+```css
+.valid {
+  background: lightgreen;
+}
 
+.invalid {
+  background: pink;
+}
+```
 
+So with our form we can track input into the fields and provide real-time validation feedback.  Next we want to actually **do** something with the data, but our `NewStudentForm` component doesn't, and shouldn't need to care what happens with the submitted data.
+
+**Question**:  Looking at our Ada student list app to date, what component would be managing the student data?
+
+## Handling Submissions
+
+Now we want to handle when the user submits the form.  We can add a function as an event handler.
+
+```javascript
+// new_student_form.js
+...
+handleFormSubmit = (event) => {
+  event.preventDefault();
+  const newStudent = {
+    name: this.state.name,
+    email: this.state.email,
+  };
+
+  this.setState({
+    name: '',
+    email: '',
+  });
+
+  // Now we need to do something with the student...
+}
+```
+
+And in `render` we add an event handler for the submit action to the `form`.
+
+```javascript
+// new_student_form.js
+...
+<form className="new-student-form" onSubmit={this.handleFormSubmit}>
+...
+```
+
+So we have a way to detect submit events on the form, but no way to get data to the rest of the application.  To solve this we will modify the `StudentCollection` component to pass a function as a prop to the `NewStudentForm` so it can pass the student data back.
+
+![passing a callback through props](./images/passing-callback.png)
+
+First adding a callback function to `StudentCollection` and passing that function to `NewStudentForm` as a prop.
+
+```javascript
+// student_collection.js
+...
+// callback function to add students to the list
+addStudent = (student) => {
+    const students = this.state.students;
+
+    students.push(student);
+    this.setState({
+      students
+    });
+  }
+
+render() {
+  let studentComponents = this.state.students.map(function(student) {
+    return <Student key={ student.name } name={ student.name } email={ student.email }/>
+  });
+
+  return (
+    <div>
+      <h3>Students</h3>
+      { studentComponents }
+      <NewStudentForm addStudent={this.addStudent} />
+    </div>
+  )
+}
+...
+```
+
+Then we can update the `handleFormSubmit` function.
+
+```javascript
+// new_student_form.js
+...
+handleFormSubmit = (event) => {
+  event.preventDefault();
+  const newStudent = {
+    name: this.state.name,
+    email: this.state.email,
+  };
+
+  this.setState({
+    name: '',
+    email: '',
+  });
+
+  this.props.addStudent(student);
+}
+```
+
+Now if we test the app we should now be able to add students to the list.  However there are some issues.
+1.  We can add students with no names
+2.  We can add students with invalid email addresses
+
+**Exercise**  Update the app to prevent form submission if the name is blank, or the email field is invalid.
+
+## Final Note: Prop Validation
```

Comment:
I think prop validation is super valuable but it seems like maybe it should be a whole separate discussion. This very brief mention of it might bring up more questions than answers.